api/envoy/extensions/transport_sockets/tls/v3/tls.proto

```diff
@@ -275,6 +275,12 @@ message CommonTlsContext {
   // [#not-implemented-hide:]
   CertificateProviderPluginInstance tls_certificate_provider_instance = 14;
 
+  // Custom TLS context provider.
+  //
+  // Select TLS context based on TLS client hello.
+  // If empty, defaults to native TLS context selection behavior.
+  config.core.v3.TypedExtensionConfig custom_tls_context_provider = 16;
```

Comment:
Please rename/document that this is for selecting a TLS certificate. This is user-facing documentation, and it isn't clear what a `TLS context` is outside of the implementation.

Reply:
Updated the comment. how about use `custom_tls_certificate_selector` instead of `custom_tls_context_provider`?

Reply:
After a deeper thought, I think `custom_tls_certificate_provider` could be better?
Here is my thoughts:
1. an extension could provide a certificate that not from `tls_certificates` or `tls_certificate_sds_secret_configs`, not just selecting from them, so, `provider` might be better.
2. TLS context is internel implementation, so, `tls_certificate` might be better.

Reply:
Are we going to allow this code to select a certificate that currently doesn't exist in the configuration (static or dynamic)? If so, that leads to a lot of open questions about how it will be implemented.

I agree that `context` shouldn't be in the name, as it is internal.

I think `custom_tls_certificate_selector` is still clearer to the user than `provider`.

Reply:
> Are we going to allow this code to select a certificate that currently doesn't exist in the configuration (static or dynamic)? If so, that leads to a lot of open questions about how it will be implemented.

yes.
1. an extension could get certificates in its own way, i.e. read from local file system, or remote server.
2. a certificate should be init to `SSL_CTX`.
3. we need an API to init `SSL_CTX` by using the common TLS configurations, i.e. certificate validation and etc. This could be implemented in another PR.

> 
> I agree that `context` shouldn't be in the name, as it is internal.
> 
> I think `custom_tls_certificate_selector` is still clearer to the user than `provider`.

okay, we could use `custom_tls_certificate_selector` even an extension may get certificates in other ways, right?
I think it does make sense, there must be a selector even certificates come from other ways.
I'll rename it soon. Thanks~